### PR TITLE
Added ability to mark a card as done

### PR DESCRIPTION
**Pull request by TehThanos**
* Resolves: #534 
* Target version: master 

### Summary
This adds the ability to mark a card as done.
When the card is marked as done if it has a due date then it makes it un-editable, if it doesn't it simply displays "No due date"
Marking a card as done also hides it from the upcoming section.

### TODO

- [x] Add Rest API documentation
- [x] Design changes requested form @nimishavijay 
- [x] Remove [l10n changes](https://github.com/nextcloud/deck/pull/4137#discussion_r1000169153)
- [x] lint issues
- [x] style lint issues
- [x] document headline number wrong [here](https://github.com/nextcloud/deck/pull/4137/files#diff-43a3c216412e058983668ebe1bf88d8c7967b5d16fdc19884122d28bb9eb03e3R57)
- [ ] missing cypress tests
- [x] Move from boolean column to datetime to store that exact timestamp when a card was marked as done
  - [x] Change migration
  - [x] Properly handle the datetime in the backend
  - [x] Maybe indicate the done timestamp in the frontend on hover of the icon or in the details view next to the done label
  - [x] Expose it through the caldav backend https://github.com/nextcloud/deck/pull/4137#issuecomment-1521730474
  - [x] Double check visibility of icons in archived state
- [x] Find a proper layout for modal/sidebar for the done state and buttons

## Follow up issues
- [ ] Done by to store which user marked it (as suggested by @z0rgster)

### Checklist

- [x] Code is properly formatted
- [x] Sign-off message is added to all commits
- [ ] Tests (unit, integration, api and/or acceptance) are included
- [x] Documentation (manuals or wiki) has been updated or is not required

### Screenshots

<details>
<summary>Current state (12th September)</summary>

Screenshot 2023-09-12 at 15 42 26

Screenshot 2023-09-12 at 15 42 32

Screenshot 2023-09-12 at 15 42 35

</details>

### CODE REVIEW

**Comment by juliushaertl** [comment]
> I ran the tests and the only ones failing are the ones in this file tests/integration/features/acl.feature.
> The weird thing is they fail regardless of if I include my commits or not, is this expected or do I have a server configuration error?

What exact failure do you see there?

File: lib/Service/CardService.php
Comment (juliushaertl) [inline lib/Service/CardService.php:560]:
```suggestion
	public function done(int $id): Card {
```

File: lib/Service/CardService.php
Comment (juliushaertl) [inline lib/Service/CardService.php:287]:
```suggestion
	public function update($id, $title, $stackId, $type, $owner, $description = '', $order = 0, $duedate = null, $deletedAt = null, $archived = null, ?bool $done = null) {
```

Comment (juliushaertl) [reply lib/Service/CardService.php:287]: We can use type hints these days, so would be nice to at least already add them for new arguments or new methods :)

File: src/components/cards/badges/Done.vue
Comment (juliushaertl) [inline src/components/cards/badges/Done.vue:75]: Seems to still target the due date ;)

File: src/store/card.js
Comment (juliushaertl) [inline src/store/card.js:332]:
```suggestion
			const updatedCard = card.done ? await apiClient.markCardAsUndone(card) : await apiClient. markCardAsDone(card)
```

File: lib/Controller/CardApiController.php
Comment (juliushaertl) [inline lib/Controller/CardApiController.php:97]:
```suggestion
	public function update($title, $type, $owner, $description = '', $order = 0, $duedate = null, $archived = null, ?bool $done = null) {
```

File: lib/Controller/CardController.php
Comment (juliushaertl) [inline lib/Controller/CardController.php:142]: @stefan-niedermann Would you prefer to have those two methods also for the REST API or do you think an optional parameter on the cards PUT method is enough? 

Also we should at least increase the API version (minor) then so that the clients can detect if the new done parameter is already present.

**Review by juliushaertl — COMMENTED** [review]
Very nice, really cool to see this pull request :) I've left a few comments inline, but it looks good to me in general.

Maybe you could post some screenshots of the frontend changes as well in order to get a review from our designers @nimishavijay and @jancborchardt :)

File: lib/Controller/CardController.php
Comment (stefan-niedermann) [reply lib/Controller/CardController.php:142]: @desperateCoder what do you prefer?

Comment (stefan-niedermann) [reply lib/Controller/CardController.php:142]: @juliushaertl how will react old Deck versions to the optional parameter? Will they just ignore it or throw an error?

Comment (juliushaertl) [reply lib/Controller/CardController.php:142]: It would just be ignored

File: src/components/cards/badges/Done.vue
Comment (TehThanos) [reply src/components/cards/badges/Done.vue]: Oups sorry 😅 Fixed in [63b6c08] https://github.com/nextcloud/deck/pull/4137/commits/63b6c0821c367af9efa868f27b22d0ae6806a4f2)

File: lib/Service/CardService.php
Comment (TehThanos) [reply lib/Service/CardService.php]: Uh okay I was just following the rest of the codebase. Fixed along with all the other type hits in d24bdc1

File: lib/Controller/CardApiController.php
Comment (TehThanos) [reply lib/Controller/CardApiController.php]: Fixed in d24bdc1

File: lib/Service/CardService.php
Comment (TehThanos) [reply lib/Service/CardService.php]: Fixed in d24bdc1

File: lib/Controller/CardController.php
Comment (desperateCoder) [reply lib/Controller/CardController.php]: Then I'd prefer the optional parameter in the PUTs.  it would be a non-breaking change and everyone is happy

File: lib/Migration/Version1000Date20200306161713.php
Comment (TehThanos) [reply lib/Migration/Version1000Date20200306161713.php]: I used 10900 as a version number and I also rolled back the changes, please let me know if it's okay.

File: lib/Migration/Version10900Date20221020074643.php
Comment (juliushaertl) [inline lib/Migration/Version10900Date20221020074643.php:6]: Make sure to add you to the copyright header ;)

File: lib/Controller/CardController.php
Comment (TehThanos) [reply lib/Controller/CardController.php]: The `CardService` also has `archived` but the `CardController` doesn't should I add it or just pass down null?

**Comment by TehThanos** [comment]
> > I ran the tests and the only ones failing are the ones in this file tests/integration/features/acl.feature.
> > The weird thing is they fail regardless of if I include my commits or not, is this expected or do I have a server configuration error?
> 
> What exact failure do you see there?

Remaking the whole environment after pruning everything seems to have fixed it, not sure what was causing it sorry for the trouble.

File: lib/Controller/CardController.php
Comment (juliushaertl) [reply lib/Controller/CardController.php]: For the API only the CardApiController is relevant, so that is fine. We currently have this separation between the API that web uses and the one that is accessible for external apps. In the long term it would be good to migrate those to a common OCS API so that we don't have this duplication anymore, but that is something out of scope for this PR.

File: lib/Migration/Version1000Date20200306161713.php
Comment (juliushaertl) [reply lib/Migration/Version1000Date20200306161713.php]: Perfect, thanks for the quick adjustment :)

**Comment by TehThanos** [comment]
> Very nice, really cool to see this pull request :) I've left a few comments inline, but it looks good to me in general.
> 
> Maybe you could post some screenshots of the frontend changes as well in order to get a review from our designers @nimishavijay and @jancborchardt :)

Thank you! 
<details>
  <summary>Here are some images</summary>
  
  
bigcard_dark_done_nodue
card_dark_done
sidebar_dark_done_due
sidebar_dark_notdone_due
sidebard_light_done_nodue
sidebar_light_done_due
sidebar_light_notdone_nodue
</details>

**Comment by TehThanos** [comment]
Uh also the tests are currently broken so sorry for that, I'm gonna push a fix in a few minutes.

**Comment by TehThanos** [comment]
I am not sure what's up with my dev environment but now when I try to run the integration tests I get `APM: Locking APM for 300 seconds for reason: Cannot connect to the agent` and while the do run, it takes forever. I don't have time today so I haven't run them but I'll take a look at it tomorrow, sorry.

**Comment by TehThanos** [comment]
`composer run psalm` still returns `ERROR: MoreSpecificImplementedParamType - lib/Migration/Version10900Date20221020074643.php:44:56` and I am not sure how to fix it since I also get `UndefinedClass` for `SimpleMigrationStep` on my system. If anyone can help me I would appreciate it. Please also take a closer look at commit 44fbb5a6

File: lib/Migration/Version10900Date20221020074643.php
Comment (marcelklehr) [inline lib/Migration/Version10900Date20221020074643.php:40]: To fix the psalm error, you can change this line
```suggestion
	 * @param Closure() $schemaClosure
```

File: src/components/card/CardSidebarTabDetails.vue
Comment (marcelklehr) [inline src/components/card/CardSidebarTabDetails.vue:128]: I think this is a bit weird, UX wise. I would expect this to show "Done", even if no due date was set. Or "Done without due date" cc @nextcloud/designers

Comment (marcelklehr) [reply src/components/card/CardSidebarTabDetails.vue:128]: My bad, I can't read. 
image

File: src/components/cards/badges/Done.vue
Comment (marcelklehr) [inline src/components/cards/badges/Done.vue:26]: What's the status on these icon classes for nc 25? They don't seem to show for me.

Comment (marcelklehr) [reply src/components/cards/badges/Done.vue:26]: cc @juliushaertl

Comment (juliushaertl) [reply src/components/cards/badges/Done.vue:26]: Yes, those would need to be migrated to use https://www.npmjs.com/package/vue-material-design-icons

The checkmark icon is already used in https://github.com/nextcloud/deck/blob/0c59a85c9c221ca930044a9831894e663a4cd8af/src/components/cards/CardBadges.vue#L36

File: lib/Migration/Version10900Date20221020074643.php
Comment (juliushaertl) [inline lib/Migration/Version10900Date20221020074643.php:56]: Best practice would be to only return the schema if there was a change:

```suggestion
		$table = $schema->getTable('deck_cards');
		if (!$table->hasColumn('done')) {
			$table->addColumn('done', 'boolean', [
				'default' => false,
				'notnull' => true,
			]);
			
			return $schema;
		}
		
		return null;
```

**Comment by nuokh** [comment]
Is there anything blocking this PR? Looks like all completed checks passed.

**Comment by TehThanos** [comment]
> Is there anything blocking this PR? Looks like all completed checks passed.

Yes [this](https://github.com/nextcloud/deck/pull/4137#discussion_r1059360111) and [this](https://github.com/nextcloud/deck/pull/4137#discussion_r1059422560)
I'm sorry, but I haven't had the time to fix them yet. I'm hoping that I will be able to before May ends.

**Comment by raimund-schluessler** [comment]
Since Deck offers an exposure of its cards via CalDAV, I think it would make sense to align with the RFC 5545. Specifically, the done/completed state of a todo is tracked by the `COMPLETED` property which stores a `Date-Time` the todo was completed at. See https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.2.1.

So instead of storing a `boolean` value `done` in the database, I would propose to call the property `completed` and let it store a date with UTC time instead. That would better map to the properties of a todo defined in RFC 5545.

File: lib/Db/Card.php
Comment (raimund-schluessler) [inline lib/Db/Card.php:110]:
```suggestion
		$this->addType('completed', 'datetime');
```
See the reasoning for this in https://github.com/nextcloud/deck/pull/4137#issuecomment-1521730474. Of course, all other occurrences of `done` would need to be adjusted as well.

**Comment by nimishavijay** [comment]
Just took a look at this! Very nice feature! Just a few points of feedback for the design:

- [ ] In the action menu: the "Mark card as done" can be moved up to the second spot or even first. I assume if a person uses this feature, they would use it very often, and having it right next to the "Delete" button increases the risk of dangerous misclicks. Also, if it is moved up, then it keeps "Archive" and "Delete" next to each other, which is expected.
- [ ] Inside the card in sidebar/modal view: 
  - [ ] The archive button could be a tertiary button (So a button without a background)
  - [ ] The done indication next to the due date can be pill shaped so it looks like a tag and use the variable `--color-success`
  - [ ] There can be some differentiation between the done and not done state. 
    - [ ] "Mark as done" could be a secondary button 
    - [ ] "Mark as not done" could be a tertiary button and change wording "Mark as not done" to "Done"
- [ ] In the columns: Since the card is already quite small, I suggest the indication of the done state in the card be communicated though just an icon instead of an icon and text. We can use the [check circle icon](https://fonts.google.com/icons?selected=Material%20Icons%3Acheck_circle%3A) in the color `--color-success`

It looks very good already, nice work! :)

File: src/components/card/CardSidebarTabDetails.vue
Comment (raimund-schluessler) [inline src/components/card/CardSidebarTabDetails.vue:28]: These should be `NcButton` components, see https://nextcloud-vue-components.netlify.app/#/Components/NcButton

File: src/components/card/CardSidebarTabDetails.vue
Comment (raimund-schluessler) [inline src/components/card/CardSidebarTabDetails.vue:121]: Since this actions menu only has one action, it should simply be an `NcButton` as well.

Comment (raimund-schluessler) [reply src/components/card/CardSidebarTabDetails.vue:121]: And use a proper material design icon, see the `NcButton` docs.

**Comment by luka-nextcloud** [comment]
To recap:

- [ ] fix comment https://github.com/nextcloud/deck/pull/4137#discussion_r1059360111
- [ ] fix comment https://github.com/nextcloud/deck/pull/4137#discussion_r1059422560
- [ ] adjust design according to comment https://github.com/nextcloud/deck/pull/4137#issuecomment-1521988946
- [ ] fix comment https://github.com/nextcloud/deck/pull/4137#pullrequestreview-1400569133
- [ ] fix comment https://github.com/nextcloud/deck/pull/4137#pullrequestreview-1400570439
- [ ] add cypress tests
- [ ] fix merge conflicts

**Comment by TehThanos** [comment]
And https://github.com/nextcloud/deck/pull/4137#issuecomment-1521730474 which is actually the only thing I haven't done yet, if I remember correctly. I haven't pushed yet so as not to spam the issue, but I could if you want.

**Comment by juliushaertl** [comment]
If you have a more recent state maybe you can rebase your changes and push, we were trying to see what is missing to get some old PRs like this merged, but if you are still up for continuing that would be very welcome of course :)

**Comment by TehThanos** [comment]
Sorry but I haven't run tests or linting yet because I need to set up my environment again, I'll do it sometime tomorrow.

**Comment by juliushaertl** [comment]
No worries, let me and @luka-nextcloud have another review round on the changes since last time then.

**Comment by TehThanos** [comment]
Turns out there where a lot of wrong things about the previous commit, sorry about that.
Unfortunately `make test` exits with an error with the only message being `Parse error: syntax error, unexpected token "match", expecting identifier in phar:///var/www/html/apps-extra/deck/build/tools/phpunit.phar/phpunit/Framework/MockObject/Builder/Match.php on line 15`

File: docs/User_documentation_en.md
Comment (Flix6x) [inline docs/User_documentation_en.md:64]: The task~s~ is

File: docs/API.md
Comment (Flix6x) [inline docs/API.md:635]: If -> Whether

File: docs/API.md
Comment (Flix6x) [inline docs/API.md:636]: If -> Whether

**Review by Flix6x — COMMENTED** [review]
Small linguistic suggestions in my review. (Really looking forward to this feature, thank you for putting in the work!)

**Comment by luka-nextcloud** [comment]
Thanks for your update, here are a list of things that are remaining:
- [ ] lint issues
- [ ] style lint issues
- [ ] document headline number wrong [here](https://github.com/nextcloud/deck/pull/4137/files#diff-43a3c216412e058983668ebe1bf88d8c7967b5d16fdc19884122d28bb9eb03e3R57)
- [ ] missing cypress tests

**Comment by raimund-schluessler** [comment]
> Thanks for your update, here are a list of things that are remaining:
> 
> - [ ] lint issues
> 
> - [ ] style lint issues
> 
> - [ ] document headline number wrong [here](https://github.com/nextcloud/deck/pull/4137/files#diff-43a3c216412e058983668ebe1bf88d8c7967b5d16fdc19884122d28bb9eb03e3R57)
> 
> - [ ] missing cypress tests

I would say https://github.com/nextcloud/deck/pull/4137#issuecomment-1521730474 also deserves some consideration.

File: lib/Db/Card.php
Comment (juliushaertl) [reply lib/Db/Card.php:110]: Yes, i think that makes a lot of sense. Not only for the dav backend but also for future features.

Added as a todo entry to the first post: 
> - [ ]  Move from boolean column to datetime to store that exact timestamp when a card was marked as done

**Comment by juliushaertl** [comment]
I'll try to collect open topics now in the first post to make the conversation a bit cleaner.

File: l10n/en_GB.js
Comment (juliushaertl) [reply l10n/en_GB.js]: This one is still valid

File: lib/Migration/Version10900Date20221020074643.php
Comment (juliushaertl) [reply lib/Migration/Version10900Date20221020074643.php]: One addition that we would need is to also change the type in the original migration, so that new installations can skip the later one.

**Comment by juliushaertl** [comment]
Note you can ignore the one cypress failure, this is also an issue on the main branch

**Comment by juliushaertl** [comment]
Thanks again for the nice PR. I did a rebase onto the latest main branch, i'd pick this up to adapt to storing the datetime instead of a boolean during the next days so we can get this in soonish. Hope that works for you :)

**Comment by TehThanos** [comment]
Yes, I'm very sorry I haven't been more active.

**Comment by juliushaertl** [comment]
Moved to a datetime now and rebased again. Will put a checklist on what is still missing above

**Review by juliushaertl — APPROVED** [review]

**Comment by stefan-niedermann** [comment]
So how does this feature eventually work?

- The REST endpoints are the same as before?
- The card properties do have an additional timestamp property from `Done`?
- Is there an UI toggle for the `Done` property or a automatism when moving from / to the last list? Or both?
- Are `ETag`s updated properly when the flag changes?

**Comment by donty** [comment]
Seen the update for v 1.11.1 come through on our Nextcloud but I don't see the Done option in the UI anywhere logical - to me anyhow ;-)  Looks like it was merged before the update so just wondering if I missed something?

**Comment by shoetten** [comment]
> Seen the update for v 1.11.1 come through on our Nextcloud but I don't see the Done option in the UI anywhere logical - to me anyhow ;-) Looks like it was merged before the update so just wondering if I missed something?

It was merged before, but not backported to the v1.11 branch ;) See the release notes. It will ship with `v1.12.0`, probably with the next major of nextcloud, i guess.

**Comment by donty** [comment]
Thanks!  So I did miss something ;-)

**Comment by sickstee** [comment]
Would it be possible to implement marking a card as done in the "reminders" app on IOS or any other to-do app that pulls infos via CalDav from the server? By now, marking a card as done in the "reminders" app doesn't pull the state to the deck app :octocat:

**Comment by stefan-niedermann** [comment]
@TehThanos and @juliushaertl  would you mind having a look at [some open questions](https://github.com/nextcloud/deck/pull/4137#issuecomment-1802155486) to be able to ship this feature also for the Deck Android app? 🙂

**Comment by TehThanos** [comment]
> The REST endpoints are the same as before?

Yes, with the addition of `/cards/[id]/done` and `/cards/[id]/undone`
> The card properties do have an additional timestamp property from Done?

The `done` property is a `datetime` if the card isn't done it will be `null`
> Is there an UI toggle for the Done property or a automatism when moving from / to the last list? Or both?

I'm not sure what exactly you mean, there is a button to mark the card as done or not done, I'm don't know about any automations.

> Are ETags updated properly when the flag changes?

I believe so.

**Comment by abelbabel** [comment]
> > Seen the update for v 1.11.1 come through on our Nextcloud but I don't see the Done option in the UI anywhere logical - to me anyhow ;-) Looks like it was merged before the update so just wondering if I missed something?
> 
> It was merged before, but not backported to the v1.11 / [stable27](https://github.com/nextcloud/deck/tree/stable27) branch ;) See the release notes. It will ship with `v1.12.0`, probably with the next major of nextcloud, i guess.

Thank you for this feature!

I use Nextcloud 28.0.1.1 with Deck 1.12 and PHP 8.1 and cannot see the "Mark done"-option/-button ...

Did I miss something?

**Pull request merged**
